services/upgrader/upgrader.go

```diff
@@ -0,0 +1,179 @@
+package upgrader
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"errors"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"runtime"
+	"sync/atomic"
+
+	"github.com/mattermost/mattermost-server/v5/model"
+)
+
+var upgradePercentage int64
+var upgradeError error
+var upgrading int32
+
+type writeCounter struct {
+	total  int64
+	readed int64
+}
+
+func (wc *writeCounter) Write(p []byte) (int, error) {
+	n := len(p)
+	wc.readed += int64(n)
+	percentage := (wc.readed * 100) / wc.total
+	if percentage == 0 {
+		upgradePercentage = 1
+	} else if percentage == 100 {
+		upgradePercentage = 99
+	} else {
+		upgradePercentage = percentage
+	}
+	return n, nil
+}
+
+func getUpgradeDirectory() (string, error) {
+	exePath, err := os.Executable()
+	if err != nil {
+		return "", err
+	}
+	return filepath.Dir(exePath), nil
+}
+
+func getCurrentVersionTgzUrl() string {
+	return "https://releases.mattermost.com/" + model.CurrentVersion + "/mattermost-" + model.CurrentVersion + "-linux-amd64.tar.gz"
+}
+
+func canIUpgrade() error {
+	if runtime.GOARCH != "amd64" {
+		return errors.New("Upgrades only supported for amd64 architectures.")
+	}
+	if runtime.GOOS != "linux" {
+		return errors.New("Upgrades only supported for linux operating systems.")
+	}
+	return nil
+}
+
+func canIUpgradeToE0() error {
+	if err := canIUpgrade(); err != nil {
+		return err
+	}
+	if model.BuildEnterpriseReady == "true" {
+		return errors.New("You can't upgrade your code to enterprise because you are already in enterprise code")
+	}
+	return nil
+}
+
+func UpgradeToE0() error {
+	if !atomic.CompareAndSwapInt32(&upgrading, 0, 1) {
+		return errors.New("One upgrade is already running.")
+	}
+	defer atomic.CompareAndSwapInt32(&upgrading, 1, 0)
+
+	upgradePercentage = 1
+	upgradeError = nil
+
+	if err := canIUpgradeToE0(); err != nil {
+		upgradePercentage = 0
+		upgradeError = err
+		return err
+	}
+	filename, err := download(getCurrentVersionTgzUrl())
```

Comment:
If I understand correctly, 0-99 percentage is the download and the rest to 100 is the unzip/creating the final binary.  I guess that is an ok approx. , given that network download time normally takes the most, maybe in the future we can refine the info we give to the user and let him know that upgrade time is actually 99% download time.

Reply:
Yes, it is more about provide a more or less accurate percentage while he is waiting, but I don't see much value on adding correctness here, or extra info here, because is just a "progress bar/percentage of completion" that is not so important. It is better to have it correct, or to let the user know what is exactly going on, yes, but removing effort from other places to put it here, I don't see it as the best way to invest the effort.

Reply:
Agree this is not a priority, it is more that it is a misleading message, if we call this upgrade progress and 99% is download (e.g. if it breaks, it might be important for them to know where it broke - in download or when deploying the binaries - they are different and should be handled differently.)

Reply:
Actually, the correct error is reported as part of the status if there was an error. so this give you the information about what was the error about. In this case the error is about downloading, and if it fails extracting the binary is going to faile with the proper error.